### ISSUE TITLE
Fix ConfirmTransaction

### ISSUE THREAD

**Issue by rom1504** [PrismarineJS/mineflayer#1811]
https://github.com/PrismarineJS/mineflayer/blob/master/lib/plugins/inventory.js#L329 the loop does not check if the queue is empty

However to fix it it's needed to understand why this function is built that way. I don't really understand what it's supposed to do.

**Comment by u9g** [comment]
what would there be to confirm if the windowClickQueue is empty? The whole point of confirming a click is to make sure that the server allows clicking there. So If you are getting a click confirmation without clicking first, then how is the server recieving a click?

**Comment by kirillbul** [comment]
> what would there be to confirm if the windowClickQueue is empty? The whole point of confirming a click is to make sure that the server allows clicking there. So If you are getting a click confirmation without clicking first, then how is the server recieving a click?

On some servers, an error is generated during the connection, but the bot logs in and crashes.
The bot enters, but an windowClickQueue is empty error appears and the bot turns off

**Comment by u9g** [comment]
> The bot enters, but an windowClickQueue is empty error appears and the bot turns off

I don't really understand the way you said this, so how does a vanilla client handle this?

should we just return if the length of the array is 0?

**Comment by rom1504** [comment]
> what would there be to confirm if the windowClickQueue is empty? The whole point of confirming a click is to make sure that the server allows clicking there. So If you are getting a click confirmation without clicking first, then how is the server recieving a click?

Read the code, we loop and remove elements from the queue but never check if it's become empty
So it will fail

You're saying it's about confirming clicks but I don't get what the code is trying to do

**Comment by u9g** [comment]
We loop and confirm every click till we get to the desired click, because if we've confirmed a click past what we have in the queue, clearly the rest of the clicks before it were successful, this then allows us to actually do whatever the transaction packet told us (accept/reject the transaction)

ref: https://wiki.vg/Protocol#Window_Confirmation_.28clientbound.29

**Comment by rom1504** [comment]
> because if we've confirmed a click past what we have in the queue, clearly the rest of the clicks before it were successful

why?
it might be true for vanilla, but not for buggy spigot plugins

**Comment by u9g** [comment]
> > because if we've confirmed a click past what we have in the queue, clearly the rest of the clicks before it were successful
> 
> why?
> it might be true for vanilla, but not for buggy spigot plugins

I didn't write it, I'm just trying to fill in the blanks for you. Of course if you want perfect support, you look at how the vanilla client does it

**Comment by rom1504** [comment]
yes the point of this issue is to fix this confirmTransaction function to work for most cases or at least not crash
currently it crashes for several uses (it tries to find id in undefined click during that loop)
